File: network.py
import numpy as np

def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

X = np.array([[0,0],
              [0,1],
              [1,0],
              [1,1]])

y = np.array([[0], [1], [1], [0]])

input_neurons = 2
hidden_neurons = 2
output_neurons = 1
learning_rate = 0.5
epochs = 10000

np.random.seed(42)
w_input_hidden = np.random.uniform(-1, 1, (input_neurons, hidden_neurons))
w_hidden_output = np.random.uniform(-1, 1, (hidden_neurons, output_neurons))
b_hidden = np.random.uniform(-1, 1, (1, hidden_neurons))
b_output = np.random.uniform(-1, 1, (1, output_neurons))

for epoch in range(epochs):
    # Forward pass
    hidden_input = np.dot(X, w_input_hidden) + b_hidden
    hidden_output = sigmoid(hidden_input)

    final_input = np.dot(hidden_output, w_hidden_output) + b_output
    final_output = sigmoid(final_input)

    # Compute error
    error = y - final_output

    # Backpropagation
    delta_output = error * sigmoid_derivative(final_output)
    delta_hidden = delta_output.dot(w_hidden_output.T) * sigmoid_derivative(hidden_output)

    # Update weights and biases
    w_hidden_output += hidden_output.T.dot(delta_output) * learning_rate
    b_output += np.sum(delta_output, axis=0, keepdims=True) * learning_rate
    w_input_hidden += X.T.dot(delta_hidden) * learning_rate
    b_hidden += np.sum(delta_hidden, axis=0, keepdims=True) * learning_rate

print("Neural Network Predictions:")
hidden_output = sigmoid(np.dot(X, w_input_hidden) + b_hidden)
final_output = sigmoid(np.dot(hidden_output, w_hidden_output) + b_output)
for i, x in enumerate(X):
    print(f"Input: {x} -> Output: {final_output[i][0]:.4f}")

File: network.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define INPUT_NEURONS 2
#define HIDDEN_NEURONS 2
#define OUTPUT_NEURONS 1
#define TRAINING_SAMPLES 4
#define LEARNING_RATE 0.5
#define EPOCHS 10000

double sigmoid(double x) {
    return 1.0 / (1.0 + exp(-x));
}

double sigmoid_derivative(double x) {
    return x * (1.0 - x);
}

double rand_weight() {
    return ((double)rand() / RAND_MAX) * 2 - 1;
}

int main() {
    srand(time(NULL));

    double X[TRAINING_SAMPLES][INPUT_NEURONS] = {
        {0, 0},
        {0, 1},
        {1, 0},
        {1, 1}
    };
    double y[TRAINING_SAMPLES] = {0, 1, 1, 0};

    double w_input_hidden[INPUT_NEURONS][HIDDEN_NEURONS];
    double w_hidden_output[HIDDEN_NEURONS];
    double b_hidden[HIDDEN_NEURONS];
    double b_output;

    for (int i = 0; i < INPUT_NEURONS; i++)
        for (int j = 0; j < HIDDEN_NEURONS; j++)
            w_input_hidden[i][j] = rand_weight();

    for (int j = 0; j < HIDDEN_NEURONS; j++) {
        w_hidden_output[j] = rand_weight();
        b_hidden[j] = rand_weight();
    }
    b_output = rand_weight();

    for (int epoch = 0; epoch < EPOCHS; epoch++) {
        for (int sample = 0; sample < TRAINING_SAMPLES; sample++) {
            // Forward pass
            double hidden[HIDDEN_NEURONS];
            for (int j = 0; j < HIDDEN_NEURONS; j++) {
                hidden[j] = b_hidden[j];
                for (int i = 0; i < INPUT_NEURONS; i++)
                    hidden[j] += X[sample][i] * w_input_hidden[i][j];
                hidden[j] = sigmoid(hidden[j]);
            }

            double output = b_output;
            for (int j = 0; j < HIDDEN_NEURONS; j++)
                output += hidden[j] * w_hidden_output[j];
            output = sigmoid(output);

            double error = y[sample] - output;

            double delta_output = error * sigmoid_derivative(output);

            double delta_hidden[HIDDEN_NEURONS];
            for (int j = 0; j < HIDDEN_NEURONS; j++)
                delta_hidden[j] = delta_output * w_hidden_output[j] * sigmoid_derivative(hidden[j]);

            for (int j = 0; j < HIDDEN_NEURONS; j++)
                w_hidden_output[j] += LEARNING_RATE * delta_output * hidden[j];
            b_output += LEARNING_RATE * delta_output;

            for (int i = 0; i < INPUT_NEURONS; i++)
                for (int j = 0; j < HIDDEN_NEURONS; j++)
                    w_input_hidden[i][j] += LEARNING_RATE * delta_hidden[j] * X[sample][i];
            for (int j = 0; j < HIDDEN_NEURONS; j++)
                b_hidden[j] += LEARNING_RATE * delta_hidden[j];
        }
    }

    printf("Predictions:\n");
    for (int sample = 0; sample < TRAINING_SAMPLES; sample++) {
        double hidden[HIDDEN_NEURONS];
        for (int j = 0; j < HIDDEN_NEURONS; j++) {
            hidden[j] = b_hidden[j];
            for (int i = 0; i < INPUT_NEURONS; i++)
                hidden[j] += X[sample][i] * w_input_hidden[i][j];
            hidden[j] = sigmoid(hidden[j]);
        }

        double output = b_output;
        for (int j = 0; j < HIDDEN_NEURONS; j++)
            output += hidden[j] * w_hidden_output[j];
        output = sigmoid(output);

        printf("Input: [%g, %g] -> Output: %g\n", X[sample][0], X[sample][1], output);
    }

    return 0;
}
